source/dofs/dof_handler_policy.cc

```diff
@@ -3641,6 +3641,9 @@ namespace internal
           const parallel::distributed::Triangulation< dim, spacedim > *triangulation
             = (dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>*>
                (&dof_handler.get_triangulation()));
+          Assert (triangulation != nullptr,
+                  ExcMessage("The function communicate_dof_indices_on_marked_cells() "
+                             "only works with parallel triangulations."));
```

Comment:
Two comments here:
1. Should we also allow `parallel::shared::tria` here?
2. I'm not sure that this will silence the coverity warning. I would prefer if we wrote something like
```cpp
if (const auto *triangulation = 
    dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim>*>(&dof_handler.get_triangulation()))
  {
    const int ierr = MPI_Barrier(triangulation->get_communicator());
    AssertThrowMPI(ierr);
  }
else
  {
    Assert(false, ExcMessage("The function communicate_dof_indices_on_marked_cells() "
                             "only works with parallel distributed triangulations."));
  }
```
I think every static analyzer will be happier with that.

Reply:
It's probably best to be as restrictive as possible and only allow `p::d::Triangulation` here.